### ISSUE TITLE
use blk_account_io_done instead of blk_account_io_completion

### ISSUE THREAD

**Issue by gurugio** [iovisor/bcc#2091]
https://github.com/iovisor/bcc/blob/6dc8ec89ef71425af52a57f3ebbaadc9f77931ca/tools/biolatency.py#L122

biolatency tool attaches kprobe at blk_account_io_completion() function to check when a request is finished.
But sometimes request can be finished partially and blk_account_io_completion() is called again.
I think blk_account_io_done() is called when a request is finally finished and biolatency tool should attach blk_account_io_done(), not blk_account_io_completion().

**Comment by yonghong-song** [comment]
Let me check with experts on this. In latest kernel, we have the following code,
```
bool blk_update_request(struct request *req, blk_status_t error,
                unsigned int nr_bytes)
{       
        int total_bytes;
                        
        trace_block_rq_complete(req, blk_status_to_errno(error), nr_bytes);
        
        if (!req->bio)
                return false;
        
        if (unlikely(error && !blk_rq_is_passthrough(req) &&
                     !(req->rq_flags & RQF_QUIET)))
                print_req_error(req, error);
                        
        blk_account_io_completion(req, nr_bytes);

        total_bytes = 0;
        while (req->bio) {
                struct bio *bio = req->bio;
                unsigned bio_bytes = min(bio->bi_iter.bi_size, nr_bytes);

                if (bio_bytes == bio->bi_iter.bi_size)
                        req->bio = bio->bi_next;
                
                /* Completion has already been traced */
                bio_clear_flag(bio, BIO_TRACE_COMPLETION);
                req_bio_endio(req, bio, bio_bytes, error);

                total_bytes += bio_bytes;
                nr_bytes -= bio_bytes;

                if (!nr_bytes)
                        break;
        }
```

So kprobe on `blk_account_io_completion` seems aligning with kernel tracepoint.

**Comment by yonghong-song** [comment]
@gurugio chatted with a few internal fs experts. yes, sometimes request will be finished partially and `blk_account_io_completion()` will be called multiple times. this happens for loop device and some old scsi. For accounting perspective, `blk_account_io_done()` is the right place to track the final finishing line of the result.
Do you mind to send a pull request for this change? Thanks!

**Comment by brendangregg** [comment]
Thanks @gurugio and @yonghong-song.

We can update the tools, but we should also check again if we can switch to tracepoints instead, so that we don't have to keep updating them as the kernel changes. Many of these tools were written when we only had kprobe support in BPF, but now we have tracepoints (plus the new raw tracepoints).

See tcplife.py for how I supported both kprobes and tracepoints with bpf_text_kprobe & bpf_text_tracepoint. That makes it easiest to maintain I think.

**Comment by yonghong-song** [comment]
@brendangregg Make sense, kernel block layer does have some trace points and we should look at them as they are more stable.

**Comment by gurugio** [comment]
I will try to make a patch after finishing some work ;-)

**Comment by gurugio** [comment]
I checked kernel 4.4 and 4.13, and found some block events indicating starting/finishing IO.
(https://www.ibm.com/developerworks/community/blogs/58e72888-6340-46ac-b488-d31aa4058e9c/entry/exploring_the_linux_storage_path_tracing_block_i_o_kernel_events?lang=en)
block_bio_queue and block_bio_complete for bio, and block_rq_insert and block_rq_complete for request, look like starting/finishing events.

I think we can use them.

**Comment by yonghong-song** [comment]
Agreed that `block_bio_queue/block_bio_complete` and `block_rq_insert/block_rq_complete` tracepoints should work. Note that for `block_rq_complete`, `req->bio` indicates whether this is a partial request completion or not. `req->bio == NULL` indicates the last request done. in bcc bpf programs, we may want to ignore all the partial requests with `reg->bio != NULL`.

**Comment by gurugio** [comment]
Good point. Thank you.

**Comment by gurugio** [comment]
I created a pull-request #2109.

So far, block_rq_insert/complete tracepoint does not provide the pointer to the request.
I cannot check how much time each request takes.

```
root@ws00837:/sys/kernel/debug/tracing/events# cat block/block_rq_issue/format 
name: block_rq_issue
ID: 1044
format:
	field:unsigned short common_type;	offset:0;	size:2;	signed:0;
	field:unsigned char common_flags;	offset:2;	size:1;	signed:0;
	field:unsigned char common_preempt_count;	offset:3;	size:1;	signed:0;
	field:int common_pid;	offset:4;	size:4;	signed:1;

	field:dev_t dev;	offset:8;	size:4;	signed:0;
	field:sector_t sector;	offset:16;	size:8;	signed:0;
	field:unsigned int nr_sector;	offset:24;	size:4;	signed:0;
	field:unsigned int bytes;	offset:28;	size:4;	signed:0;
	field:char rwbs[8];	offset:32;	size:8;	signed:1;
	field:char comm[16];	offset:40;	size:16;	signed:1;
	field:__data_loc char[] cmd;	offset:56;	size:4;	signed:1;

print fmt: "%d,%d %s %u (%s) %llu + %u [%s]", ((unsigned int) ((REC->dev) >> 20)), ((unsigned int) ((REC->dev) & ((1U << 20) - 1))), REC->rwbs, REC->bytes, __get_str(cmd), (unsigned long long)REC->sector, REC->nr_sector, REC->comm
```

I will contact block-layer developers and ask for adding more information including the pointer to request and bio structure.

**Comment by yonghong-song** [comment]
pull request #2109 looks good. We can live it for now. One way to get the request pointer is to use raw_tracepoint, which is invented for this exact reason. Look at `tools/runqslower.py` as an example. But raw_tracepoint is only available at 4.17 and later, so your change is good anyway and will merge it after the test.
